src/main/java/BinaryTree/AVLtree.java

```diff
@@ -0,0 +1,344 @@
+                }
+            } else if (current.getLeftNode() != null && current.getRightNode() == null) {
+                if (current.getParentNode() == null) { // removing a root node
+                    current.getLeftNode().setParentNode(null);
+                    current = current.getLeftNode();
+                } else {
+                    if (current.getParentNode().getKey() < current.getRightNode().getKey()) {
+                        current.getParentNode().setRightNode(current.getLeftNode());
+                        current.getLeftNode().setParentNode(current.getParentNode());
+                    } else {
+                        current.getParentNode().setLeftNode(current.getLeftNode());
+                        current.getLeftNode().setParentNode(current.getParentNode());
+                    }
+                }
+            }
+            // Deleting a node with two children
+            else {
+                BinaryNode<T> successor = find_successor();
+                bst_delete(successor.getKey());
+                current.setKey(successor.getKey());
+            }
+        }
+        updateHeight(current);
+    }
+
+    public BinaryNode<T> find_successor() {
+        BinaryNode<T> original_position = current;
+        moveToRightNode();
+        while (current.getLeftNode() != null) {
+            moveToLeftNode();
+        }
+        BinaryNode<T> successor = current;
+        current = original_position;
+        return successor;
+    }
+
+    public void delete(int delete_key) {
+        // perform standard bst deletion
+        bst_delete(delete_key);
+
+        // balance the tree
+        int depth_difference = getDepthDifference(current);
+        if (depth_difference > 1 && getDepthDifference(current.getLeftNode()) >= 0) { // Left Left case
+            rightRotation();
+        } else if (depth_difference < -1 && getDepthDifference(current.getRightNode()) <= 0) { // Right Right case
+            leftRotation();
+        } else if (depth_difference > 1 && getDepthDifference(current.getLeftNode()) < 0) { //Left Right case
+            moveToLeftNode();
+            leftRotation();
+            moveToParentNode();
+            rightRotation();
+        } else if (depth_difference < -1 && getDepthDifference(current.getRightNode()) > 0) { // Right Left case
+            moveToRightNode();
+            rightRotation();
+            moveToParentNode();
+            leftRotation();
+        }
+    }
+
+    public void rightRotation() {
+        BinaryNode<T> newRoot = current.getLeftNode();
+        BinaryNode<T> moveSubtreeToRight = new BinaryNode<T>();
+
+        if (newRoot.getRightNode() != null) {
+            moveSubtreeToRight = newRoot.getRightNode();
+        }
+
+        newRoot.setRightNode(current);
+        newRoot.setParentNode(current.getParentNode());
+        current.setParentNode(newRoot);
+        current.setLeftNode(moveSubtreeToRight);
+
+        if (moveSubtreeToRight != null) {
```

Comment:
will this ever be null?

Reply:
yes in                  5                              3
                        3      7       ->           2      5
                      2                              1           7
                     1
 3 does not have right child to be moved to 5's left node

Reply:
you are setting `moveSubtreeToRight` at line 215.
How can this be not null